{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Async Connection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create an output file with the current date in its name \n",
    "import datetime\n",
    "output_file = f\"streaming_data/streaming_data_{datetime.datetime.now().strftime('%Y-%m-%d_%H-%M-%S')}.txt\"\n",
    "\n",
    "import logging\n",
    "import asyncio\n",
    "from azure.eventhub.aio import EventHubConsumerClient\n",
    "\n",
    "import nest_asyncio # to allow for event loop in jupyter notebook\n",
    "nest_asyncio.apply()\n",
    "\n",
    "logger = logging.getLogger(\"azure.eventhub\")\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "#Credentials\n",
    "EVENT_HUB_CONN_STR='Endpoint=sb://factored-datathon.servicebus.windows.net/;SharedAccessKeyName=datathon_group_1;SharedAccessKey=2GETvVt0FxyM0bo0Qau4inlmC/w3t4Uut+AEhAnAEgk=;EntityPath=factored_datathon_amazon_reviews_1'\n",
    "EVENTHUB_NAME=\"factored_datathon_amazon_reviews_1\"\n",
    "CONSUMER_GROUP=\"datacticos\"\n",
    "\n",
    "#Connection to Event Hub\n",
    "async def on_event(partition_context, event):\n",
    "    #logger.info(f\"Received event from partition {partition_context.partition_id} with sequence number {event.sequence_number} and data:\\n {list(event.body)}\")\n",
    "    #print(f\"Received event from partition {partition_context.partition_id} with sequence number {event.sequence_number} and data:\\n {list(event.body)}\")\n",
    "    \n",
    "    with open(output_file, \"a+\") as file:\n",
    "        file.write(f\"{list(event.body)}\\n\")\n",
    "\n",
    "    await partition_context.update_checkpoint(event)\n",
    "\n",
    "async def receive():\n",
    "    client =  EventHubConsumerClient.from_connection_string(conn_str=EVENT_HUB_CONN_STR,consumer_group=CONSUMER_GROUP,eventhub_name=EVENTHUB_NAME,)\n",
    "    async with client:\n",
    "        await client.receive(\n",
    "            on_event=on_event,\n",
    "            starting_position=\"-1\",  # \"-1\" is from the beginning of the partition.\n",
    "        )\n",
    "        # receive events from specified partition:\n",
    "        # await client.receive(on_event=on_event, partition_id='0')\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    \n",
    "    with open(output_file, \"w\"):\n",
    "            pass  # Create or clear the output file before starting the process\n",
    "\n",
    "    loop = asyncio.get_event_loop()\n",
    "    loop.run_until_complete(receive())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Test Formating data to json\n",
    "\n",
    "output_file=\"output_file_4.json\"\n",
    "\n",
    "import datetime\n",
    "import json\n",
    "import logging\n",
    "import asyncio\n",
    "from azure.eventhub.aio import EventHubConsumerClient\n",
    "\n",
    "import nest_asyncio # to allow for event loop in jupyter notebook\n",
    "nest_asyncio.apply()\n",
    "\n",
    "logger = logging.getLogger(\"azure.eventhub\")\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "#Credentials\n",
    "EVENT_HUB_CONN_STR='Endpoint=sb://factored-datathon.servicebus.windows.net/;SharedAccessKeyName=datathon_group_1;SharedAccessKey=2GETvVt0FxyM0bo0Qau4inlmC/w3t4Uut+AEhAnAEgk=;EntityPath=factored_datathon_amazon_reviews_1'\n",
    "EVENTHUB_NAME=\"factored_datathon_amazon_reviews_1\"\n",
    "CONSUMER_GROUP=\"datacticos\"\n",
    "\n",
    "#Connection to Event Hub\n",
    "async def on_event(partition_context, event):\n",
    "    #logger.info(f\"Received event from partition {partition_context.partition_id} with sequence number {event.sequence_number} and data:\\n {list(event.body)}\")\n",
    "    #print(f\"Received event from partition {partition_context.partition_id} with sequence number {event.sequence_number} and data:\\n {list(event.body)}\")\n",
    "    \n",
    "    with open(output_file, \"a+\") as file:\n",
    "        data_str=list(event.body)[0].decode('utf-8')\n",
    "        #convert string to json\n",
    "        data=json.loads(data_str)\n",
    "        json.dump(data, file)\n",
    "        file.write(\"\\n\")\n",
    "\n",
    "    await partition_context.update_checkpoint(event)\n",
    "\n",
    "async def receive():\n",
    "    client =  EventHubConsumerClient.from_connection_string(conn_str=EVENT_HUB_CONN_STR,consumer_group=CONSUMER_GROUP,eventhub_name=EVENTHUB_NAME,)\n",
    "    async with client:\n",
    "        await client.receive(\n",
    "            on_event=on_event,\n",
    "            starting_position=\"-1\",  # \"-1\" is from the beginning of the partition.\n",
    "        )\n",
    "        # receive events from specified partition:\n",
    "        # await client.receive(on_event=on_event, partition_id='0')\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    \n",
    "    with open(output_file, \"w\"):\n",
    "            pass  # Create or clear the output file before starting the process\n",
    "\n",
    "    loop = asyncio.get_event_loop()\n",
    "    loop.run_until_complete(receive())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:azure.eventhub.aio._eventprocessor.event_processor:EventProcessor 'd6e492d0-f04c-4d73-9d2b-9726dae23af6' is being started\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: None -> <ConnectionState.START: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.START: 0> -> <ConnectionState.HDR_SENT: 2>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.HDR_SENT: 2> -> <ConnectionState.HDR_SENT: 2>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.HDR_SENT: 2> -> <ConnectionState.OPEN_PIPE: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.UNMAPPED: 0> -> <SessionState.BEGIN_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPEN_PIPE: 4> -> <ConnectionState.OPEN_SENT: 7>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPEN_SENT: 7> -> <ConnectionState.OPENED: 9>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.BEGIN_SENT: 1> -> <SessionState.MAPPED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._cbs_async:CBS completed opening with status: <ManagementOpenResult.OK: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACH_SENT: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACH_SENT: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACH_SENT: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACH_SENT: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACH_SENT: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACH_SENT: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACH_SENT: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACH_SENT: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.MAPPED: 3> -> <SessionState.END_SENT: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPENED: 9> -> <ConnectionState.CLOSE_SENT: 11>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.CLOSE_SENT: 11> -> <ConnectionState.END: 13>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.END_SENT: 4> -> <SessionState.DISCARDING: 6>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACH_SENT: 4> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.DETACH_SENT: 4> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACH_SENT: 4> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.DETACH_SENT: 4> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACH_SENT: 4> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.DETACH_SENT: 4> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACH_SENT: 4> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.DETACH_SENT: 4> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub.aio._eventprocessor.event_processor:EventProcessor 'd6e492d0-f04c-4d73-9d2b-9726dae23af6' has claimed partition '0'\n",
      "INFO:azure.eventhub.aio._eventprocessor.event_processor:start ownership '0', checkpoint None\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: None -> <ConnectionState.START: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.START: 0> -> <ConnectionState.HDR_SENT: 2>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.HDR_SENT: 2> -> <ConnectionState.HDR_SENT: 2>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.HDR_SENT: 2> -> <ConnectionState.OPEN_PIPE: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.UNMAPPED: 0> -> <SessionState.BEGIN_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPEN_PIPE: 4> -> <ConnectionState.OPEN_SENT: 7>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPEN_SENT: 7> -> <ConnectionState.OPENED: 9>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.BEGIN_SENT: 1> -> <SessionState.MAPPED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._cbs_async:CBS completed opening with status: <ManagementOpenResult.OK: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[2], line 56\u001b[0m\n\u001b[0;32m     53\u001b[0m         \u001b[39mpass\u001b[39;00m  \u001b[39m# Create or clear the output file before starting the process\u001b[39;00m\n\u001b[0;32m     55\u001b[0m loop \u001b[39m=\u001b[39m asyncio\u001b[39m.\u001b[39mget_event_loop()\n\u001b[1;32m---> 56\u001b[0m loop\u001b[39m.\u001b[39;49mrun_until_complete(receive())\n",
      "File \u001b[1;32mc:\\Users\\GABRIEL\\Desktop\\Datathon\\Code\\datathon\\Lib\\site-packages\\nest_asyncio.py:84\u001b[0m, in \u001b[0;36m_patch_loop.<locals>.run_until_complete\u001b[1;34m(self, future)\u001b[0m\n\u001b[0;32m     82\u001b[0m     f\u001b[39m.\u001b[39m_log_destroy_pending \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m\n\u001b[0;32m     83\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mnot\u001b[39;00m f\u001b[39m.\u001b[39mdone():\n\u001b[1;32m---> 84\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_run_once()\n\u001b[0;32m     85\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_stopping:\n\u001b[0;32m     86\u001b[0m         \u001b[39mbreak\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\GABRIEL\\Desktop\\Datathon\\Code\\datathon\\Lib\\site-packages\\nest_asyncio.py:107\u001b[0m, in \u001b[0;36m_patch_loop.<locals>._run_once\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    100\u001b[0m     heappop(scheduled)\n\u001b[0;32m    102\u001b[0m timeout \u001b[39m=\u001b[39m (\n\u001b[0;32m    103\u001b[0m     \u001b[39m0\u001b[39m \u001b[39mif\u001b[39;00m ready \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_stopping\n\u001b[0;32m    104\u001b[0m     \u001b[39melse\u001b[39;00m \u001b[39mmin\u001b[39m(\u001b[39mmax\u001b[39m(\n\u001b[0;32m    105\u001b[0m         scheduled[\u001b[39m0\u001b[39m]\u001b[39m.\u001b[39m_when \u001b[39m-\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtime(), \u001b[39m0\u001b[39m), \u001b[39m86400\u001b[39m) \u001b[39mif\u001b[39;00m scheduled\n\u001b[0;32m    106\u001b[0m     \u001b[39melse\u001b[39;00m \u001b[39mNone\u001b[39;00m)\n\u001b[1;32m--> 107\u001b[0m event_list \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_selector\u001b[39m.\u001b[39;49mselect(timeout)\n\u001b[0;32m    108\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_process_events(event_list)\n\u001b[0;32m    110\u001b[0m end_time \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtime() \u001b[39m+\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_clock_resolution\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\selectors.py:323\u001b[0m, in \u001b[0;36mSelectSelector.select\u001b[1;34m(self, timeout)\u001b[0m\n\u001b[0;32m    321\u001b[0m ready \u001b[39m=\u001b[39m []\n\u001b[0;32m    322\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 323\u001b[0m     r, w, _ \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_select(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_readers, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_writers, [], timeout)\n\u001b[0;32m    324\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mInterruptedError\u001b[39;00m:\n\u001b[0;32m    325\u001b[0m     \u001b[39mreturn\u001b[39;00m ready\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\selectors.py:314\u001b[0m, in \u001b[0;36mSelectSelector._select\u001b[1;34m(self, r, w, _, timeout)\u001b[0m\n\u001b[0;32m    313\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_select\u001b[39m(\u001b[39mself\u001b[39m, r, w, _, timeout\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[1;32m--> 314\u001b[0m     r, w, x \u001b[39m=\u001b[39m select\u001b[39m.\u001b[39mselect(r, w, w, timeout)\n\u001b[0;32m    315\u001b[0m     \u001b[39mreturn\u001b[39;00m r, w \u001b[39m+\u001b[39m x, []\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:azure.eventhub._pyamqp.aio._cbs_async:Token will expire soon - attempting to refresh.\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:An error occurred when detaching the link: AMQPConnectionError('Error condition: ErrorCondition.SocketError\\n Error Description: Can not read frame due to exception: [WinError 10053] Se ha anulado una conexión establecida por el software en su equipo host')\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:An error occurred when detaching the link: AMQPConnectionError('Error condition: ErrorCondition.SocketError\\n Error Description: Can not read frame due to exception: [WinError 10053] Se ha anulado una conexión establecida por el software en su equipo host')\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:An error occurred when detaching the link: AMQPConnectionError('Error condition: ErrorCondition.SocketError\\n Error Description: Can not read frame due to exception: [WinError 10053] Se ha anulado una conexión establecida por el software en su equipo host')\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:An error occurred when ending the session: AMQPConnectionError('Error condition: ErrorCondition.SocketError\\n Error Description: Can not read frame due to exception: [WinError 10053] Se ha anulado una conexión establecida por el software en su equipo host')\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.MAPPED: 3> -> <SessionState.UNMAPPED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:An error occurred when closing the connection: AMQPConnectionError('Error condition: ErrorCondition.SocketError\\n Error Description: Can not read frame due to exception: [WinError 10053] Se ha anulado una conexión establecida por el software en su equipo host')\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPENED: 9> -> <ConnectionState.END: 13>\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: None -> <ConnectionState.START: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.START: 0> -> <ConnectionState.HDR_SENT: 2>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.HDR_SENT: 2> -> <ConnectionState.HDR_SENT: 2>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.HDR_SENT: 2> -> <ConnectionState.OPEN_PIPE: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.UNMAPPED: 0> -> <SessionState.BEGIN_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPEN_PIPE: 4> -> <ConnectionState.OPEN_SENT: 7>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPEN_SENT: 7> -> <ConnectionState.OPENED: 9>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.BEGIN_SENT: 1> -> <SessionState.MAPPED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._cbs_async:CBS completed opening with status: <ManagementOpenResult.OK: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:An error occurred when detaching the link: AMQPConnectionError('Error condition: ErrorCondition.SocketError\\n Error Description: Can not read frame due to exception: [WinError 10053] Se ha anulado una conexión establecida por el software en su equipo host')\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:An error occurred when detaching the link: AMQPConnectionError('Error condition: ErrorCondition.SocketError\\n Error Description: Can not read frame due to exception: [WinError 10053] Se ha anulado una conexión establecida por el software en su equipo host')\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:An error occurred when detaching the link: AMQPConnectionError('Error condition: ErrorCondition.SocketError\\n Error Description: Can not read frame due to exception: [WinError 10053] Se ha anulado una conexión establecida por el software en su equipo host')\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.ATTACHED: 3> -> <LinkState.DETACHED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:An error occurred when ending the session: AMQPConnectionError('Error condition: ErrorCondition.SocketError\\n Error Description: Can not read frame due to exception: [WinError 10053] Se ha anulado una conexión establecida por el software en su equipo host')\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.MAPPED: 3> -> <SessionState.UNMAPPED: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:An error occurred when closing the connection: AMQPConnectionError('Error condition: ErrorCondition.SocketError\\n Error Description: Can not read frame due to exception: [WinError 10053] Se ha anulado una conexión establecida por el software en su equipo host')\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPENED: 9> -> <ConnectionState.END: 13>\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._transport_async:Transport connect failed: OSError('failed to resolve broker hostname')\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: None -> <ConnectionState.START: 0>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.START: 0> -> <ConnectionState.HDR_SENT: 2>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.HDR_SENT: 2> -> <ConnectionState.HDR_SENT: 2>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.HDR_SENT: 2> -> <ConnectionState.OPEN_PIPE: 4>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.UNMAPPED: 0> -> <SessionState.BEGIN_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPEN_PIPE: 4> -> <ConnectionState.OPEN_SENT: 7>\n",
      "INFO:azure.eventhub._pyamqp.aio._connection_async:Connection state changed: <ConnectionState.OPEN_SENT: 7> -> <ConnectionState.OPENED: 9>\n",
      "INFO:azure.eventhub._pyamqp.aio._session_async:Session state changed: <SessionState.BEGIN_SENT: 1> -> <SessionState.MAPPED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link receiver state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._management_link_async:Management link sender state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._cbs_async:CBS completed opening with status: <ManagementOpenResult.OK: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.DETACHED: 0> -> <LinkState.ATTACH_SENT: 1>\n",
      "INFO:azure.eventhub._pyamqp.aio._link_async:Link state changed: <LinkState.ATTACH_SENT: 1> -> <LinkState.ATTACHED: 3>\n",
      "INFO:azure.eventhub._pyamqp.aio._cbs_async:Token will expire soon - attempting to refresh.\n",
      "INFO:azure.eventhub._pyamqp.aio._cbs_async:Token will expire soon - attempting to refresh.\n"
     ]
    }
   ],
   "source": [
    "#Saving data as json\n",
    "\n",
    "import datetime\n",
    "\n",
    "#output_file=\"output_file_4.json\"\n",
    "\n",
    "output_file = f\"streaming_data/streaming_data_{datetime.datetime.now().strftime('%Y-%m-%d_%H-%M-%S')}.json\"\n",
    "\n",
    "\n",
    "import json\n",
    "import logging\n",
    "import asyncio\n",
    "from azure.eventhub.aio import EventHubConsumerClient\n",
    "\n",
    "import nest_asyncio # to allow for event loop in jupyter notebook\n",
    "nest_asyncio.apply()\n",
    "\n",
    "logger = logging.getLogger(\"azure.eventhub\")\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "#Credentials\n",
    "EVENT_HUB_CONN_STR='Endpoint=sb://factored-datathon.servicebus.windows.net/;SharedAccessKeyName=datathon_group_1;SharedAccessKey=2GETvVt0FxyM0bo0Qau4inlmC/w3t4Uut+AEhAnAEgk=;EntityPath=factored_datathon_amazon_reviews_1'\n",
    "EVENTHUB_NAME=\"factored_datathon_amazon_reviews_1\"\n",
    "CONSUMER_GROUP=\"datacticos\"\n",
    "\n",
    "#Connection to Event Hub\n",
    "async def on_event(partition_context, event):\n",
    "    #logger.info(f\"Received event from partition {partition_context.partition_id} with sequence number {event.sequence_number} and data:\\n {list(event.body)}\")\n",
    "    #print(f\"Received event from partition {partition_context.partition_id} with sequence number {event.sequence_number} and data:\\n {list(event.body)}\")\n",
    "    \n",
    "    with open(output_file, \"a+\") as file:\n",
    "        data_str=list(event.body)[0].decode('utf-8')\n",
    "        #convert string to json\n",
    "        data=json.loads(data_str)\n",
    "        json.dump(data, file)\n",
    "        file.write(\"\\n\")\n",
    "\n",
    "    await partition_context.update_checkpoint(event)\n",
    "\n",
    "async def receive():\n",
    "    client =  EventHubConsumerClient.from_connection_string(conn_str=EVENT_HUB_CONN_STR,consumer_group=CONSUMER_GROUP,eventhub_name=EVENTHUB_NAME,)\n",
    "    async with client:\n",
    "        await client.receive(\n",
    "            on_event=on_event,\n",
    "            starting_position=\"-1\",  # \"-1\" is from the beginning of the partition.\n",
    "        )\n",
    "        # receive events from specified partition:\n",
    "        # await client.receive(on_event=on_event, partition_id='0')\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    \n",
    "    with open(output_file, \"w\"):\n",
    "            pass  # Create or clear the output file before starting the process\n",
    "\n",
    "    loop = asyncio.get_event_loop()\n",
    "    loop.run_until_complete(receive())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "datathon",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
